### PR TITLE
SPU LLVM WIP

### PR DESCRIPTION
**Pull request by Nekotekina**

### CODE REVIEW

**Comment by Zangetsu38** [comment]
crash in SPU asmjit/llvm for tlou not fixed

**Comment by Asinin3** [comment]
Thanks for addressing the issue i found in https://github.com/RPCS3/rpcs3/pull/4816/commits/3459ef160b10a6dda4e12e903f8b881b8e777aeb F1 Race stars now works with SPU LLVM.
neko_2018-06-26_1479

**Comment by kerojin180** [comment]
this game looks cool whats its name?

**Comment by kerojin180** [comment]
Good job you guys are amazing!

**Comment by kerojin180** [comment]
#4798 have a similar issue is this related to it?

**Comment by Asinin3** [comment]
@kerojin180 Github isn't the place for discussion, and you can edit your own posts instead of double posting or deleting comments then making a new one. There are many games that don't work with SPU LLVM, addressing one issue isn't going to fix everything. I can confirm that Persona 5 still doesn't work with SPU LLVM even with this PR.

**Comment by allens39** [comment]
Tiger Woods 12 Masters will now not run with spu llvm and crashes so bad you have to kill the process manually in the task manager...
fatal
 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/2138129/RPCS3.log)

**Comment by Xcedf** [comment]
Tested few games:
among the working ones 
GoW3, Uncharted 2, Killzone 2, Ninja Gaiden 2 are seriously faster than ASMJit, yet GoW3 and Ninja Gaiden are suffering from major graphic issues
Still an outstanding work with new recompiler, Thanks a lot

**Comment by Crimsom13** [comment]
minecraft now crashes instantly instead of loading into game level.

**Comment by kerojin180** [comment]
@M220 is it really working for you?

**Comment by Zangetsu38** [comment]
@kerojin180  Yes, works is normal now, it same thing with my build on llvm 7. but just here is very wip, final llvm 7 released on september 2018

**Comment by kerojin180** [comment]
RPCS3 still crashes on my ryzen 3 2200g sadly ...

**Comment by kerojin180** [comment]
@Zangetsu38 what are your specs?

**Comment by Zangetsu38** [comment]
Ryzen 2700X

**Comment by kerojin180** [comment]
Ah what version of rpcs3 ? 7018 or 7017?

**Comment by Zangetsu38** [comment]
[](https://cdn.discordapp.com/attachments/272875751773306881/461632774177161229/unknown.png)

**Comment by M220** [comment]
RDR Tested ✔😀
A little faster then LLVM 6 and a faw graphical issues fixed but Have Major Graphical Issues that makes the game pretty hard to play :D♡

**Comment by maximilian578** [comment]
SEINS;GATE 0 [BLJM61303] now works on spu llvm.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2142971/RPCS3.log.gz)

**Comment by kerojin180** [comment]
It's working on persona 5 now
Fps improvement is very noticable

When there is crowd Fps takes a big hit like always but I have my cup to blame for that not good enough hehe else hits 30fps

**Comment by YaPeL** [comment]
intel haswell 4690k, stock speed, non tsx trickery enable, around a 20-30% speed bump, not perfect, but its nearly there, amazing job, and I was losing faith on my shitty processor.

**Comment by Parthorisian** [comment]
Experimental LLVMRecompiler can run Demon's Souls **[BLUS30443]** like master, but sound is much better
Conan **[BLUS30069]** - also runs as usual
Shadows of the Damned [BLUS30653] - same situation, but sometimes **Access violation reading location** like Conan, previous game (UnrealEngine3)
Dante Inferno **[BLUS30405]** - still major graphical glithes https://i.imgur.com/gNUbrAe.png
Wet **[BLUS30403]** - some more glitches - Rubi can disappear for performin some acrobatic moves (in movies too)+some textures partially rendered https://i.imgur.com/wdcmfnP.png
But for all games have been presented some weird freezes+sound delay sometimes (comparing to asmjit)

**Comment by raven02** [comment]
DBZ : Burst of Limit also start working with SPU LLVM

**Comment by tabnk** [comment]
Ni no Kuni: Wrath of the White Witch crash/stuck on load

**Comment by Asinin3** [comment]
@tabnk Unrelated to this PR, please don't spam. https://github.com/RPCS3/rpcs3/issues/4589

**Comment by Kravickas** [comment]
**NHL 15**
ASMJIT Safe
```
U {SPU[0x2000004] Thread (Job Manager - Job Thread) [0x0011c]} SPU TODO: [0x1b4] Unexpected instruction on $SP: SELB
U {SPU[0x2000004] Thread (Job Manager - Job Thread) [0x0015c]} SPU TODO: [0x204] Unexpected instruction on $SP: SELB
U {SPU[0x2000000] Thread (Job Manager - Job Thread) [0x001fc]} SPU TODO: [0x2a4] Unexpected instruction on $SP: SELB
U {SPU[0x2000005] Thread (Job Manager - Job Thread) [0x00184]} SPU TODO: [0x1b8] Unexpected instruction on $SP: SELB
U {SPU[0x2000000] Thread (Job Manager - Job Thread) [0x01f98]} SPU TODO: [0x21cc] Unexpected instruction on $SP: SELB
```
Idk if ''TODO: **[0x21cc]** Unexpected'' make difference
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2145445/RPCS3.log.gz)

**Comment by Nekotekina** [comment]
If it's a bunch of SELB, it's safe.

**Comment by Emulator-Team-2** [comment]
**DYNASTY WARRIORS: Strikeforce Demo**
```
W {SPU[0x2000002] Thread (CellSpursKernel2) [0x15eb0]} SPU: [0x15eec] Unexpected instruction on $SP: ORI
W {SPU[0x2000002] Thread (CellSpursKernel2) [0x16060]} SPU: [0x160a0] Unexpected instruction on $SP: ORI
W {SPU[0x2000001] Thread (CellSpursKernel1) [0x0d188]} SPU: [0xd18c] Unexpected instruction on $SP: A

```
Log:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2145917/RPCS3.log.gz)

--------------------------------------------------------------------------------------------------------------
**Fist of the North Star: Ken's Rage 2-Demo**
**This game have a bunch of ORI and A.**
**ORI-->[365]**
**A-->[5]**
```
W {SPU[0x2000003] Thread (CellSpursKernel3) [0x09bf8]} SPU: [0x9bfc] Unexpected instruction on $SP: A
W {SPU[0x2000004] Thread (CellSpursKernel4) [0x35538]} SPU: [0x35574] Unexpected instruction on $SP: ORI
```
Log:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2145960/RPCS3.log.gz)

**Comment by Nekotekina** [comment]
These are warnings

**Comment by Dokman** [comment]
Graphics in Infamous are better and now gets 4,5 fps, before we had a 3 fps so it's very massive right now, it's a lot since when it started to work

**Comment by Dokman** [comment]
i have a 1800x and 16Gb of RAM so.... is difference for me

**Comment by yahfz** [comment]
@Dokman @vietnam13231 Stop spamming. This is not a Infamous emulator.

**Comment by Zangetsu38** [comment]
no forget also bump version in pputhread ,for ppu llvm

**Comment by LuismaSP89** [comment]
My games works Ok with this PR, but in some cases "Guilty Gear XRD rev 2 for example" music stop playing after 1 minute (more or less) while in battle.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 1]
```diff
@@ -2688,18 +2690,167 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
...
+		// TODO
+		return _spu->get_ch_value(SPU_RdInMbox);
+	}
+
+	static u32 exec_read_dec(SPUThread* _spu)
+	{
+		const u32 res = _spu->ch_dec_value - static_cast<u32>(get_timebased_time() - _spu->ch_dec_start_timestamp);
+
+		if (res > 1500)
+			std::this_thread::yield();
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp]: need to check if spu loop detection is on before `std::this_thread::yield();`

**Comment by al0xf** [comment]
So for the record Demon's Souls is fixed 👌. People can test other old regressions too.

And with SPU LLVM and Block Size: Mega performance in Demon's Souls is absolutely excellent.

File: rpcs3/Emu/Cell/SPURecompiler.cpp (hunk 1 above)
Comment (Nekotekina) [reply rpcs3/Emu/Cell/SPURecompiler.cpp]: Well, yeah, I forgot

**Comment by Luffykun007** [comment]
Graphics still missing in Uncharted: Drake's Fortune with the SPU-LLVM only!

capturar
-------------------------------------------------------------------------------------------------------------------
capturar1
-------------------------------------------------------------------------------------------------------------------
capturar2

**Comment by sufianahmed2007** [comment]
Tekken 6 and tekken tag 2 working fine in spu llvm but tekken tag 2 have colour issue which is present in past and asura wrath freezes in game when I pressed any button but booting and asura can run with joystick but when I pressed any button it freezes and soul calibur 5 not compiling cache and heavenly sword also freezes ingame

**Comment by Nekotekina** [comment]
Please stop reporting broken/missing graphics, it's known and will not be fixed in this PR.

**Comment by Parthorisian** [comment]
**MEM: Access violation reading location 0x0** issue often happening for UE3 games still posess his post [ShadowsoftheDamned_RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2152894/ShadowsoftheDamned_RPCS3.zip)

**Comment by sufianahmed2007** [comment]
Today I tested tomb raider trilogy with SPU LLVM
Tomb raider legend 
Working but graphics is not rendered correctly its same as ASMJIT but gameplay is smooth 
Tomb raider Anniversary
Working properly with SPU LLVM Like a charm no problem at all
Tomb raider Underworld
Trilogy version freezing in both SPU LLVM AND ASMJIT 
SO I TRIED STANDALONE VERSION 
IT WORKS LIKE A CHARM BUT GIVING ME WHITE SCREEN AFTER INTRO CUTSCENE IN BOTH SPU LLVM AND ASMJIT 
BUT PROGRESS IS AWESOME 
HOPE I M HELPING HERE

**Comment by 0nomat0** [comment]
Tested Demon's Souls a little with SPU LLVM.... Played for ~30 min; no crashes. I didn't experiment too much with different settings, although it wouldn't run without SPU blocks = safe.  Performance was improved from 6718 in Boletarian Palace entrance (usually one of the worst areas), but Penetrator Archstone area was very stuttery.  Identical experience with kd-11's latest PR https://github.com/RPCS3/rpcs3/pull/4844.  Both builds had occasional flashes of white or black; this build seemed to have it more often. 

3570K @ 4.5 // GTX 660 SLI (2)

**Comment by Kravickas** [comment]
obrazek
this isnt showing in this PR.

Idk if i am doing something wrong or no and u know about it.

**Comment by psennermann** [comment]
7 days ago it's been written that Ni no Kuni: Wrath of the White Witch crash/stuck on load, well now it works and goes ingame...

**Comment by AniLeo** [comment]
Yes, fixed a few days ago by 6580cce571bb685d2c6bfc5b8053dfca93170db0

**Comment by sufianahmed2007** [comment]
New build improve fps in games with SPU LLVM with SPU cache enabled in cpu settings tested with tekken 6 and tekken tag 2 and shadow of the Colossus
And shadow of the Colossus and ico working fine with SPU LLVM
I have a problem that I m playing soul calibur 4 AND IT WORKED FINE but after I win first battle my game freezes and I have to force close it with task manager it happened with both SPU LLVM AND ASMJIT can anyone help me

**Comment by raveskirza** [comment]
@sufianahmed2007 SC4 issue has nothing to do with this PR. Please dont report issues unrelated to this PR. It regressed with another PR and the last working build was 0.0.5-6831

**Comment by sufianahmed2007** [comment]
@raveskirza thanks for the reply I m reporting issues I wanted to confirm that it's not happening to me only 
Sorry for going off topic I will not do that again sorry

**Comment by LuismaSP89** [comment]
Still mute some games after 1 minute or less of play. But after some testing, no all games were affected, for example bomberman ultra or folklore works ok all the time. But other games like lollipop chainsaw or guilty gear xrd rev2 mute after some seconds of gameplay. In the case of Lollipop chainsaw the sound of the cutscenes works ok, but ingame the sound mute immediately. Strange indeed...

**Comment by kerojin180** [comment]
How do I get the experimental versions of rpcs3? I am a total noob I am
sorry :(

On 6 Jul 2018 3:00 a.m., "LuismaSP89" <notifications@github.com> wrote:

Still mute some games after 1 minute or less of play. But after some
testing, no all games were affected, for example bomberman ultra or
folklore works ok all the time. But other games like lollipop chainsaw or
guilty gear xrd rev2 mute after some seconds of gameplay. In the case of
Lollipop chainsaw the sound of the cutscenes works ok, but ingame the sound
mute immediately. Strange indeed...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/RPCS3/rpcs3/pull/4816#issuecomment-402894196>, or mute
the thread

**Comment by Crimsom13** [comment]
Went from a build a day or two ago to the build today and resistance 2 crashes pre menu again.

**Comment by Josivan88** [comment]
heavy performance decrease, games more uninstable than before:
i7 4790
1070GTX
16Gb of RAM

Red dead redemption:


[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2171459/RPCS3.log.gz)
Asura's Wrath
No sound
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2171523/RPCS3.log.gz)
Killzone 2
Not go in game and crash constantly
Fifa Street 3
Crash the emulator in "Loading pipeline objects"
TSX disabled and forced, same results.
In the crashs i can't recover RPCS3.log.gz

**Comment by Emulator-Team-2** [comment]
@Josivan88 
Don't use spu llvm in AAA title...
Broken graphic in RDR and other games is normal with spu llvm.

**Comment by Josivan88** [comment]
@Emulator-Team-2  see M220 comment, telling improvements on RDR
see Xcedf comment, telling that many AAA getting little faster, with me the opposite.

**Comment by allens39** [comment]
AAA games are years off from being playable lets face the reality here.

**Comment by allens39** [comment]
Love how @AniLeo acts like they're the boss when nekotekina and kd-11 do all the hard work. Love it!

**Comment by Josivan88** [comment]
heavy performance decrease, games more unstable than before:
all tests using SPU LLVM
i7 4790
1070GTX
16Gb of RAM

Red dead redemption:
(this PR) great FPS oscillation, but lower FPS than ASMJIT


[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2171459/RPCS3.log.gz)
on the contradiction with M220 said
Asura's Wrath
(This PR) sometimes with no sound
(Previous Master) playable
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2171523/RPCS3.log.gz)
Killzone 2
(This PR) crash constantly, even in menus
(Previous Master) stable in menus
Fifa Street 3
(This PR) Crash the emulator in "Loading pipeline objects"
(Previous Master) pass through "Loading pipeline objects"
TSX disabled and forced, same results.
In the crash i can't recover RPCS3.log.gz

**Comment by tabnk** [comment]
Virtua Fighter 5 Final Showdown

Drop performance , games unstable and hang.

**Comment by sufianahmed2007** [comment]
Soul calibur 5 now playable with SPU LLVM and there is no crashes or freezes and graphical issues at all even in my low end pc. I must say awesome improvement in development thanx developers
My specs
I5 4430
4gb ram
GTX 650 2gb

**Comment by AzaZPPL** [comment]
After this update I cannot install firmware with my CPU. Same issue as with [#4887](https://github.com/RPCS3/rpcs3/issues/4877)
AMD Phenom II x4 965 with an RX480 and 8 gigs of ram.

**Pull request merged**